### ISSUE TITLE
Is ttp:version="2" required if ttp:contentProfiles is present? 

### ISSUE THREAD

**Issue by palemieux** [w3c/ttml2#435]
TTML2 specifies that:

_a ttp:version attribute must be specified on the root tt element of a document instance if it requires support for a feature not defined by [TTML1]._

Is this a processor or document requirement, and what does "require" mean? Specifically, does the mere presence of the `ttp:contentProfiles` attribute in an otherwise TTML1 conformant document trigger the need for `ttp:version="2"`.

I do not think so, but just confirming.

**Comment by skynavga** [comment]
On Fri, Sep 29, 2017 at 10:53 AM, Pierre-Anthony Lemieux <
notifications@github.com> wrote:

> TTML2 specifies that:
>
> *a ttp:version attribute must be specified on the root tt element of a
> document instance if it requires support for a feature not defined by
> [TTML1].*
>
> Is this a processor or document requirement, and what does "require" mean?
> Specifically, does the mere presence of the ttp:contentProfiles attribute
> in an otherwise TTML1 conformant document trigger the need for
> ttp:version="2".
>
A content (document) requirement. Require means the author requires the
implementation to support the TTML2 feature. In other words, if this is
advisory, the content author could drop version, and assume that on an
implementation that doesn't support the feature (here ttp:contentProfiles),
the feature will be ignored; but, if the author wants to require support
(for ttp:contentProfiles) then the author should mark as ttp:version="2"
since it is going to require some version 2 feature.

**Comment by skynavga** [comment]
To elaborate further, if the author expects to require a TTML2 feature to
be supported, e.g., make reference (directly or indirectly) to a processor
profile that will require some TTML2 feature, then the author should
include ttp:version="2"; however, if the author does not expect to require
support for a TTML2 feature by an implementation, then they need not
include ttp:version="2".

From the processor perspective, the presence of ttp:version="2" should be
considered a warning (by the author) that support for some version 2
feature is going to be required (via a processor profile).

**Comment by palemieux** [comment]
So it looks like ttp:version="2" signals a processor requirement. In other words, the author indicates that the processor must understand one or more TTML2 features to be able to process the document accurately. Is that right?

**Comment by skynavga** [comment]
No, that isn't the intent. It is more of an advisory. Any signaling of
required support must be via a processor profile. The ttp:version is more
of a heads up that something is coming.

**Comment by palemieux** [comment]
> No, that isn't the intent. It is more of an advisory. 

The current prose uses conformance language that makes it more than an advisory:

_A ttp:version attribute must be specified on the root tt element of a document instance if it requires support for a feature not defined by [TTML1]._

It also does not capture "if the author expects to require a TTML2 feature to be supported".

I suggest the following tweak:

_A ttp:version attribute allows the author to indicate that processing of the document instance requires processing of a feature not defined by [TTML1]._

**Comment by cconcolato** [comment]
Why not explicitly refer to the use of a processor profile if the feature is required, like:

_A ttp:version attribute allows the author to indicate that the document instance uses a feature not defined by [TTML1]. Whether the processing of this feature is required or not depends on its signaling in a processor profile._

**Comment by skynavga** [comment]
Yes, that captures the intent well.

**Comment by palemieux** [comment]
@cconcolato Does the mere presence of the ttp:contentProfiles attribute trigger the presence of ttp:version, even though the document is still conformant with TTML1, i.e. does not use any other TTML2-only features? Hopefully the answer is no :)

**Comment by skynavga** [comment]
Yes, it would trigger the need for version. You would need to use
ttp:profile to avoid version.

**Comment by palemieux** [comment]
@skynavga You are not @cconcolato, but ok.

> Yes, it would trigger the need for version. You would need to use ttp:profile to avoid version.

I am not excited by that.

ttp:contentProfiles, when applied to otherwise TTML1 documents, could be a nice way of signalling multiple profiles (since ttp:profile supports only one) without implying that TTML2 feature are necessary for processing. Specifically, ttp:contentProfiles is not necessary for processing, and the mere presence of ttp:contentProfiles in a TTML1 document is permitted by TTML1 as far as I know.

**Comment by skynavga** [comment]
Let me ponder.

**Comment by cconcolato** [comment]
Note that the initial sentence that @palemieux quoted is also present in 3.3:

> In addition, it must specify a ttp:version attribute on its root tt element if it requires support for a feature not defined by [TTML1].

**Comment by palemieux** [comment]
@skynavga During the last TTWG call, you mentioned that using `ttp:version` was at the author's discretion. Can you confirm?

**Comment by skynavga** [comment]
Since it is the author who decides which features are required to be
supported, then yes it is the at the discretion of the author. Note,
however, that a process may infer that such support, i.e., V2 support, is
required even in the absence of ttp:version. For example, if the effective
processor profile requires support for a V2 feature would trigger such
inference.

**Comment by nigelmegitt** [comment]
A closely related but possibly hair-splitting question: in the text for, say, `tts:position`:
> If both `tts:position` and `tts:origin` attributes are specified on an element and `tts:position` is a supported property, then the `tts:origin` attribute must be ignored for the purpose of presentation processing.

Does "`tts:position` is a supported property" mean the same as "the `#position` feature is supported` ?

I ask because we should use consistent language for expressing support/absence of support for syntax or semantics, and since we define the feature mechanism we should probably use it.

**Comment by palemieux** [comment]
> I ask because we should use consistent language for expressing support/absence of support for syntax or semantics, and since we define the feature mechanism we should probably use it.

Why not forbid `tts:origin` if `tts:position` is present?

**Comment by nigelmegitt** [comment]
> Why not forbid `tts:origin` if `tts:position` is present?

Because then there's no fallback for TTML1 processors. The idea here is a good one IMO, that a single document can work on TTML1 processors that don't understand `tts:position` and TTML2 processors that do understand it. Essentially this provides a long term deprecation path for `tts:origin`. Of course you may think that `tts:origin` is perfectly fine and doesn't need deprecation, but then an additional translation would be needed for CSS style positioning, so in the spirit of CSS alignment moving to use of `tts:position` seems like a good idea to me.

**Comment by skynavga** [comment]
On Tue, Oct 10, 2017 at 10:49 AM, Pierre-Anthony Lemieux <
notifications@github.com> wrote:

> I ask because we should use consistent language for expressing
> support/absence of support for syntax or semantics, and since we define the
> feature mechanism we should probably use it.
>
> Why not forbid tts:origin if tts:position is present?
>
Unlike IMSC, TTML does not specify constraints on what goes into a
document. Instead, it specifies what behavior is when content may be
potentially ambiguous, as is the case if both origin and position are
specified.

**Comment by palemieux** [comment]
It looks like _[construct default processor profile](b)_ unambiguously specifies the processing semantics of `ttp:version="2"`:

_if the computed value of the ttp:version property is 2 (two), the default processor profile is the profile constructed by interning the TTML2 Presentation profile (http://www.w3.org/ns/ttml/profile/ttml2-presentation);_

Moreover, this is overriden if content or processor profiles are explicitly specified.

The following sentence can therefore be removed, and the ambiguity that led to this issue removed:

_A ttp:version attribute must be specified on the root tt element of a document instance if it requires support for a feature not defined by [TTML1]._

**Comment by skynavga** [comment]
I don't think so. The first of these quoted statements is a requirement on
processor behavior. The second is a requirement on author behavior.

Consider if the author requires support for a TTML2 feature (and specifies
ttp:version='2'), but fails to explicitly specify a content or processor
profile. In this case, the current algorithm in [construct default
processor profile] will select the dfxp-presentation or dfxp-transformation
processor profile as the default unless the document interchange context
implies an associated profile, in which case, the processor is free to
ignore all TTML2 features, including, specifically, #contentProfiles, and
not trigger processing as described under §6.1.3 (4).

On the other hand, if the author does specify ttp:version='2', but does not
explicitly specify a content or processor profile, then [construct default
processor profile] selects the ttml2-presentation or ttml2-transformation
profile unless the document interchange context implies a profile, and, as
such, triggers TTML2 (ttml2-* profile driven) behavior rather than TTML1
(dfxp-* profile driven) behavior.

In other words, if a document uses ttp:contentProfiles and does not specify
ttp:version='2', then a TTML2 processor may choose to process the document
as a TTML1 (dfxp-presentation profile) document unless the document
interchange context implies a profile.

In conclusion, if an author uses ttp:contentProfiles (and does not use
ttp:profile) and does not specify ttp:version='2', then there is no
requirement that a TTML2 processor process the ttp:contentProfiles
attribute.

Since this may be viewed as potentially ambiguous, I'm of a mind to require
that ttp:version='2' be specified in the case that

   - ttp:profile is not specified
   - either (or both) ttp:processorProfiles or (and) ttp:processorProfiles
   is (are) specified

**Comment by palemieux** [comment]
Alright... it sounds like we both agree that the current sentence (_a ttp:version attribute must be specified on the root tt element of a document instance if it requires support for a feature not defined by [TTML1]_) can yield the ambiguities.

> In other words, if a document uses ttp:contentProfiles and does not specify
ttp:version='2', then a TTML2 processor may choose to process the document
as a TTML1 (dfxp-presentation profile) document unless the document
interchange context implies a profile.

According to the _[construct effective content profile]_ algorithm, the _inferred processor profile_ can be determined based on the _effective content profile_, which can itself be determined using `ttp:contentProfiles` alone -- *regardless* of the presence of `ttp:version`.

One of the challenges here is that `ttp:version` can signal conformance to a single version of TTML, while a document can in fact conform (and be processed) by more than one version simultaneously. TTML2 moved away from the single-valued `ttp:profile` attribute specifically for this reason.

Couldn't TTML2 simply require the presence of one or more of the profile-determining attributes? An author could then explicitly signal conformance to a TTML2 profile (in addition to any other profile), e.g. `http://www.w3.org/ns/ttml/profile/ttml2-transformation`.

**Comment by skynavga** [comment]
On Sat, Oct 14, 2017 at 3:47 PM, Pierre-Anthony Lemieux <
notifications@github.com> wrote:

> Alright... it sounds like we both agree that the current sentence (*a
> ttp:version attribute must be specified on the root tt element of a
> document instance if it requires support for a feature not defined by
> [TTML1]*) can yield the ambiguities.
>
> In other words, if a document uses ttp:contentProfiles and does not specify
> ttp:version='2', then a TTML2 processor may choose to process the document
> as a TTML1 (dfxp-presentation profile) document unless the document
> interchange context implies a profile.
>
> According to the *[construct effective content profile]* algorithm, the *inferred
> processor profile* can be determined based on the *effective content
> profile*, which can itself be determined using ttp:contentProfiles alone
> -- *regardless* of the presence of ttp:version.
>
> One of the challenges here is that ttp:version can signal conformance to
> a single version of TTML,
>
That isn't quite right. ttp:version signals that an author requires
processor support for some TTML2 feature; it says nothing about content
conformance. For example, if the author requires validation support against
a specific content profile, and uses ttp:contentProfiles, a TTML2 feature,
to signal that content profile, then the author will want to specify
ttp:version='2' (or greater).

On the other hand, if the document author creates an EBU-TT-D document and
uses *only* ebuttd:conformsToStandard to signal content conformance and
does not specify ttp:version, then a TTML2 validating processor may simply
ignore validation.

> while a document can in fact conform (and be processed) by more than one
> version simultaneously. TTML2 moved away from the single-valued
> ttp:profile attribute specifically for this reason.
>
Couldn't TTML2 simply require the presence of one or more of the
> profile-determining attributes? An author could then explicitly signal
> conformance to a TTML2 profile (in addition to any other profile), e.g.
> http://www.w3.org/ns/ttml/profile/ttml2-transformation.
>
TTML does not, in general, mandate authorial behavior (though it does make
recommendations), as I seem to find myself repeating over and over again.
It defines processor behavior in the presence (or absence) of specific
content features.

See also the new PR #468 I just pushed which may alter your position.

**Comment by palemieux** [comment]
> That isn't quite right. ttp:version signals that an author requires processor support for some TTML2 feature;

That is not what the prose currently states.

The current prose *requires* `ttp:version` to be present regardless of the authors opinion.

> See also the new PR #468 I just pushed which may alter your position.

Issue #467 is unrelated.

**Comment by skynavga** [comment]
On Sat, Oct 14, 2017 at 9:52 PM, Pierre-Anthony Lemieux <
notifications@github.com> wrote:

> That isn't quite right. ttp:version signals that an author requires
> processor support for some TTML2 feature;
>
> That is not what the prose currently states.
>
> The current prose *requires* ttp:version to be present regardless of the
> authors opinion.
>
You are misreading or misinterpreting the prose.

A ttp:version attribute must be specified on the root tt element of a document
> instance
> <https://w3c.github.io/ttml2/spec/ttml2.html#terms-document-instance> if
> it requires support for a feature
> <https://w3c.github.io/ttml2/spec/ttml2.html#terms-feature> not defined
> by [TTML1] <https://w3c.github.io/ttml2/spec/ttml2.html#ttml1>.


The phrase "if it requires" means "if it [*the intended use of the document
as determined by the author*] requires support". There can be no other
interpretation, since it is only the author who determines what support is
required for a given document.

**Comment by cconcolato** [comment]
@skynavga I prefer your clarified text. It is not so clear to me that everyone will interpret it this way otherwise. Many people would assume that if you put a feature not defined in TTML1 in a TTML1 document, you want it to be supported, otherwise you wouldn't put it in the first place.

**Comment by skynavga** [comment]
On Sun, Oct 15, 2017 at 10:45 PM, Cyril Concolato <notifications@github.com>
wrote:

> @skynavga <https://github.com/skynavga> I prefer your clarified text. It
> is not so clear to me that everyone will interpret it this way otherwise.
> Many people would assume that if you put a feature not defined in TTML1 in
> a TTML1 document, you want it to be supported, otherwise you wouldn't put
> it in the first place.
>
That would be a pretty strange assumption to make if there were only TTML1
processors or one is trying to craft a document that gracefully degrades in
a TTML1 processor but demands new features in a TTML2 processor. That is,
the presence of ttp:version='2' means nothing to a TTML1 processor but
signals a TTML2 processor that at least some TTML2 feature will be required
in a TTML2 processing context. It puts the TTML2 processor on notice that
it shouldn't treat the document as a TTML1 document for one thing.

**Comment by skynavga** [comment]
To conclude my remarks on this subject
1. ttp:version is not required unless the author wishes to signal the document was authored with respect to TTML2;
2. if ttp:version is not specified, then it is considered to be defaulted to version 1; that is, a content processor should  assume that TTML1 semantics apply;
3. a TTML>=2 processor uses the resolved value of version to select between alternate behaviors that are conditioned on version; for example, a TTML2 processor that receives a TTML document lacking ttp:version may (conformantly) process the document as a TTML1 document; however, if that document designates that a TTML2 feature is required to be supported, e.g., by referring to or including a TTML profile that mandates support for a TTML2 feature, then that requirement should be satisfied or processing aborted per profile processing semantics;
4. the conclusion is: if you author a TTML document that mandates a TTML2 feature, then you should specify ttp:version='2'; otherwise, the processor is free to process it in "TTML1 Mode", i.e., according to strict TTML1 semantics.

**Comment by skynavga** [comment]
I am marking this as "editor considers closed" since I have answered the stated question.

**Comment by nigelmegitt** [comment]
@palemieux since @skynavga has answered the question in this issue, are you happy to close it?

**Comment by palemieux** [comment]
@nigelmegitt The specification does not reflect https://github.com/w3c/ttml2/issues/435#issuecomment-343685601. In particular, the specification does not state  _[in absence of ttp:profile > 2], the processor is free to process it in "TTML1 Mode", i.e., according to strict TTML1 semantics_. This conflicts with the use of `ttp:contentProfiles`. Specifically, if a TTML2 processor is free to ignore `ttp:contentProfiles` if `ttp:version` is absent, then `ttp:version` is required... but then why not use `ttp:contentProfiles` (or one of the other profiling vocabulary) instead of `ttp:version` to signal TTMLx conformance.

**Comment by skynavga** [comment]
We can't use the presence `ttp:contentProfile` or any other feature to reliably signal a specific spec version, since future versions (e.g., TTML3) will (usually) include features from prior versions.

We cannot dictate what a multi-version (and/or multi-profile) processor will do with a given document in the absence of document internal authorial declarations that the document was authored with respect to a specific version. If ttp:version is present, then it is a reliable indicator of authorial intentions about processing; however, in its absence, it will depend on specific application behavior. A specific application may (outside the specification context) decide that such a document should be processed according to a specific version, but different processors may do different things.

Of course, a profile of TTML2, say IMSC vNext, may choose to be more constrained that TTML2 in this regard. Note that we keep circling around the question of constrained design; TTML*, as such, is designed to impose minimal constraints on authors and implementers; however, profiles are free to be more constrained. You need to put your mind in the proper mode to consider this issue. They (TTML* and their profiles) have different goals.

**Comment by palemieux** [comment]
Ok, so if a document declares `ttp:processorProfile="XXXXX"` but does not declare `ttp:version="2"`, then, according to (2) at https://github.com/w3c/ttml2/issues/435#issuecomment-343685601, the processor disregards `ttp:processorProfile` and the document is processed according to a DFXP profile, right?

**Comment by palemieux** [comment]
See meeting notes at https://lists.w3.org/Archives/Public/public-tt/2017Nov/0116.html, and action https://www.w3.org/AudioVideo/TT/tracker/actions/510

**Comment by skynavga** [comment]
I have concluded my study of this question, and determined that no algorithm can reliably determine whether to use the TTML1 profile semantics or the TTML2 (or later version) profile semantics without the author providing a hint as to which version of TTML profile semantics applies.

This conclusion is based on the following assumptions:
* new versions of TTML will incorporate features defined by previous versions of TTML;
* new versions of TTML will define new features not defined by a previous version of TTML;
* new versions of TTML will define new default profiles that make reference to both pre-existing and new features;
* an author need not make explicit reference to a (content or processor) profile;
* an author may want a default profile defined by a specific version of the TTML specification to apply in the absence of an explicit profile reference.

These assumptions lead to the conclusion that the author must be provided a mechanism to signal which specification applies (in the absence of other information) so that the particular specification profile semantics are applied when determining the default profile.

Now, having said the above, I believe that we can provide an algorithm that allows the author to omit ``ttp:version`` when an explicit reference to a content profile is provided in the document. It would go something like the following:
1. add ``@version`` to the ``ttp:profile`` element, which allows the profile writer to explicitly bind a profile to a version;
2. add ``version='2'`` to the new TTML2 specific standard profiles defined in https://w3c.github.io/ttml2/spec/ttml2.html#standard-profiles;
3. enhance the algorithm for resolving effective version to work as follows
* if ``@ttp:version`` is specified, then its value is the effective version;
* otherwise, if ``ttp:contentProfiles`` is specified, then the maximum specified ``@version`` of the referenced profiles is the effective version;
* otherwise, version 1 applies.

Example 1

```
<tt ... ttp:version='2'/>
```

Example 2

```
<tt ... ttp:contentProfiles='#myProfile'>
...
<ttp:profile xml:id='myProfile' type='content' version='2'>...</ttp:profile>
...
</tt>
```

Example 3

```
<tt ... ttp:contentProfiles='http://www.w3.org/ns/ttml/profile/ttml2-presentation'/>
```

**Comment by nigelmegitt** [comment]
Why use `ttp:contentProfiles` to establish the processing semantics rather than `ttp:processorProfiles`? If we did that then the document author would be able to state which version plus profile semantics apply and then whichever the processor chooses to use defines whether TTML1 or TTML2 processor behaviour is used.

By the algorithm at https://github.com/w3c/ttml2/issues/435#issuecomment-348064332 this example:

```xml
<tt ... ttp:contentProfiles=`#ttml1Profile #ttml2profile`>
...
<ttp:profile xml:id=`ttml1Profile` type=`content` version=`1`>...</ttp:profile>
<ttp:profile xml:id=`ttml2Profile` type=`content` version=`2`>...</ttp:profile>
...
</tt>
```

would require TTML2 processing semantics even though the document is both TTML1 and TTML2 conformant. 

When deciding which TTML version to use when inferring the effective processor profile from the content profile, in the case that no processor profile is specified (as in the above example), it seems to make sense to use the _lowest_ version number rather than the highest.

The effect of doing that would be that a TTML2 processor should generate the same presentation output as a TTML1 processor for a document authored with the intention of being conformant to both TTML1 and TTML2. I _think_ this is what e.g. @mikedo asked for.

**Comment by skynavga** [comment]
@nigelmegitt I originally wrote the algorithm to include reference to ``ttp:processorProfiles``, but then removed it because it processor profile(s) is not a claim about content per se (and ``ttp:version`` was a claim about content, which implied affects on processing). However, I gave this some further thought during the night, and came to the conclusion that something needed to be said about possibly using ``ttp:processorProfiles`` as well. So I will update the proposed algorithm to do this.

As for which version applies when multiple version declarations are in scope, particularly in the case where an author can declare conformance to multiple content profiles based on different versions, it may be necessary to distinguish between effective content version and effective processor version. I will address this in the forthcoming update referred to above.

**Comment by nigelmegitt** [comment]
@skynavga that seems like a good thing to do, if we're going down this route. 

Looking at the whole problem from a different angle, the overall complexity is going up and my sense is that the likelihood of general implementations implementing that complexity interoperably is going down.

I wonder if we can align the models of TTML1 Third Edition and TTML2 sufficiently that the whole question does not arise, and draw a line in the sand that says "TTML2 processing is a superset of TTML1 Third Edition, and might differ from TTML1 Second Edition processing in some special circumstances" and leave out version identification altogether. Something to think about and discuss.

**Comment by skynavga** [comment]
@nigelmegitt yes, I've been considering the complexity angle as well; in actual implementation code, I believe it is not complex, and actually required to build multi-profile, multi-version processors; I admit that the specification details, however, make it appear rather complicated; for me, it comes down to the issue of being able to construct interoperable content and implementations, which, whether we like it or not, sometimes involves dense specification text; [consider the XSD specs as a case in point]

**Comment by cconcolato** [comment]
I think addding `@version` to the `ttp:profile` element gets things more complex. 

If the problem is to determine a version based on profile information, why not simply say that if a feature F defined in version N of the spec is required in the profile, then the version associated is the highest N across all features.

**Comment by skynavga** [comment]
On Fri, Dec 1, 2017 at 11:30 AM, Cyril Concolato <notifications@github.com>
wrote:

> I think addding @version to the ttp:profile element gets things more
> complex.
>
> If the problem is to determine a version based on profile information, why
> not simply say that if a feature F defined in version N of the spec is
> required in the profile, then the version associated is the highest N
> across all features.
>
Because V(doc) may be greater than max(V(required(F))). Consider a document
constructed for a TTML3 schema but it only requires support for features
defined by TTML1 or TTML2. See, e.g., the note that appears in
https://w3c.github.io/ttml2/spec/ttml2.html#profile-attribute-version:

*Note:*

A content processor
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-content-processor> typically
uses the declared version to perform a preliminary assessment of whether it
is capable of processing a given document instance
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-document-instance>.
However, it does not assume that the document instance
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-document-instance> actually
uses or requires support for a feature
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-feature> not defined in
prior versions. In other words, a content processor
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-content-processor> does
not reject a document instance
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-document-instance> simply
because it declares it was authored against a version of the Timed Text
Markup Language
<https://w3c.github.io/ttml2/spec/ttml2.html#terms-timed-text-markup-language>
specification
that was not yet published at the time the processor was implemented.
An important property of version that is orthogonal to feature usage or
support is an  (indirect) declaration about which schema applies, where,
different versions of the TTML schema(s) are bound to different versions of
the spec.

**Comment by skynavga** [comment]
Discussed among myself, @cconcolato and @palemieux, with tentative agreement to remove ``ttp:version`` while

* using version of specification of most recently defined required feature to determine a default schema version;

* defining new feature designators to drive selection of alternate algorithm behavior

* mandate support for dfxp profiles;

* change default profile to select profile from TTML2 profile collection
